File: backend/internal/middleware/tenant.go
package middleware

import (
	"context"
	"errors"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// TenantContext holds tenant information for the current request
type TenantContext struct {
	TenantID          string // Tenant ID from JWT claims (user's assigned tenant)
	IsSuperAdmin      bool   // Whether the user is a super admin
	EffectiveTenantID string // The tenant ID to use for queries (may differ for super-admin context switch)
}

// Context keys for tenant context
const (
	tenantContextKey       = "tenant_context"
	tenantContextHeaderKey = "X-Tenant-Context"
)

// Errors related to tenant context
var (
	ErrMissingTenantContext  = errors.New("tenant context not found")
	ErrInvalidTenantID       = errors.New("invalid tenant ID format")
	ErrTenantContextDenied   = errors.New("tenant context switch not allowed")
	ErrMissingTenantID       = errors.New("tenant ID is required")
)

// TenantContextMiddleware extracts tenant information from JWT claims and optional header
// Must be used AFTER AuthRequired middleware
func TenantContextMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		// Get user claims from context (set by AuthRequired middleware)
		userClaims, ok := GetUserClaims(c)
		if !ok {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error": "authentication required for tenant context",
			})
			return
		}

		// Create tenant context from user claims
		tenantCtx := &TenantContext{
			TenantID:          userClaims.TenantID,
			IsSuperAdmin:      userClaims.IsSuperAdmin,
			EffectiveTenantID: userClaims.TenantID, // Default to user's tenant
		}

		// Check for X-Tenant-Context header (super-admin only)
		headerTenantID := c.GetHeader(tenantContextHeaderKey)
		if headerTenantID != "" {
			// Only super-admins can use X-Tenant-Context header
			if !userClaims.IsSuperAdmin {
				c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
					"error": "tenant context switch requires super admin privileges",
					"code":  "TENANT_CONTEXT_DENIED",
				})
				return
			}

			// Validate the header tenant ID format
			if _, err := uuid.Parse(headerTenantID); err != nil {
				c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
					"error":   "invalid X-Tenant-Context header",
					"details": "tenant ID must be a valid UUID",
				})
				return
			}

			// Set effective tenant ID to the one from header
			tenantCtx.EffectiveTenantID = headerTenantID
		}

		// Store tenant context in Gin context
		c.Set(tenantContextKey, tenantCtx)

		c.Next()
	}
}

// RequireTenant is a middleware that ensures tenant context is present and valid
// Must be used AFTER TenantContextMiddleware
func RequireTenant() gin.HandlerFunc {
	return func(c *gin.Context) {
		tenantCtx, ok := GetTenantContext(c)
		if !ok {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
				"error": "tenant context required",
				"code":  "TENANT_REQUIRED",
			})
			return
		}

		// Ensure effective tenant ID is present
		if tenantCtx.EffectiveTenantID == "" {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
				"error": "tenant ID is required for this operation",
				"code":  "TENANT_ID_MISSING",
			})
			return
		}

		// Validate tenant ID format
		if _, err := uuid.Parse(tenantCtx.EffectiveTenantID); err != nil {
			c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
				"error": "invalid tenant ID format",
				"code":  "INVALID_TENANT_ID",
			})
			return
		}

		c.Next()
	}
}

// GetTenantContext retrieves the tenant context from Gin context
func GetTenantContext(c *gin.Context) (*TenantContext, bool) {
	ctx, exists := c.Get(tenantContextKey)
	if !exists {
		return nil, false
	}

	tenantCtx, ok := ctx.(*TenantContext)
	return tenantCtx, ok
}

// GetTenantIDFromGinContext extracts the effective tenant ID from Gin context
func GetTenantIDFromGinContext(c *gin.Context) (string, error) {
	tenantCtx, ok := GetTenantContext(c)
	if !ok {
		return "", ErrMissingTenantContext
	}

	if tenantCtx.EffectiveTenantID == "" {
		return "", ErrMissingTenantID
	}

	return tenantCtx.EffectiveTenantID, nil
}

// IsSuperAdminFromGinContext checks if the current user is a super admin
func IsSuperAdminFromGinContext(c *gin.Context) bool {
	tenantCtx, ok := GetTenantContext(c)
	if !ok {
		return false
	}
	return tenantCtx.IsSuperAdmin
}

// TenantContextKey is the context key type for standard context.Context
type tenantCtxKeyType string

const tenantCtxKey tenantCtxKeyType = "tenant"

// TenantInfo holds tenant information for standard Go context
type TenantInfo struct {
	TenantID     string
	IsSuperAdmin bool
}

// WithTenantContext creates a new context with tenant information
// This is used to pass tenant context to repository layer
func WithTenantContext(ctx context.Context, tenantID string, isSuperAdmin bool) context.Context {
	return context.WithValue(ctx, tenantCtxKey, &TenantInfo{
		TenantID:     tenantID,
		IsSuperAdmin: isSuperAdmin,
	})
}

// GetTenantFromContext extracts tenant information from standard Go context
// Returns tenant ID, isSuperAdmin flag, and error if not found
func GetTenantFromContext(ctx context.Context) (string, bool, error) {
	info, ok := ctx.Value(tenantCtxKey).(*TenantInfo)
	if !ok || info == nil {
		return "", false, ErrMissingTenantContext
	}
	return info.TenantID, info.IsSuperAdmin, nil
}

// GetTenantIDFromContext extracts only the tenant ID from standard Go context
func GetTenantIDFromContext(ctx context.Context) (string, error) {
	tenantID, _, err := GetTenantFromContext(ctx)
	return tenantID, err
}

// InjectTenantContext is a middleware that injects tenant context into the request's context.Context
// This makes tenant info available to repository layer via context
// Must be used AFTER TenantContextMiddleware
func InjectTenantContext() gin.HandlerFunc {
	return func(c *gin.Context) {
		tenantCtx, ok := GetTenantContext(c)
		if ok && tenantCtx.EffectiveTenantID != "" {
			// Create new context with tenant info and replace the request context
			ctx := WithTenantContext(c.Request.Context(), tenantCtx.EffectiveTenantID, tenantCtx.IsSuperAdmin)
			c.Request = c.Request.WithContext(ctx)
		}

		c.Next()
	}
}

File: ai-service/app/middleware/tenant.py
"""
Multi-Tenant Context Middleware for AI Service.

Replicates the TenantContext logic from the Go backend:
/backend/internal/middleware/tenant.go

Handles tenant isolation and supports super-admin context switching
via the X-Tenant-Context header.
"""

from dataclasses import dataclass
from typing import Optional
from uuid import UUID

from fastapi import Request, HTTPException, status

from app.middleware.auth import UserClaims


# Header key for tenant context switching (super-admin only)
TENANT_CONTEXT_HEADER = "X-Tenant-Context"


class TenantContextDeniedError(Exception):
    """Raised when a non-super-admin attempts to switch tenant context."""

    def __init__(self, message: str = "Tenant context switch requires super admin privileges"):
        self.message = message
        self.code = "TENANT_CONTEXT_DENIED"
        super().__init__(self.message)


class InvalidTenantIdError(Exception):
    """Raised when an invalid tenant ID is provided."""

    def __init__(self, message: str = "Invalid tenant ID format"):
        self.message = message
        self.code = "INVALID_TENANT_ID"
        super().__init__(self.message)


class MissingTenantContextError(Exception):
    """Raised when tenant context is missing but required."""

    def __init__(self, message: str = "Tenant context not found"):
        self.message = message
        self.code = "TENANT_REQUIRED"
        super().__init__(self.message)


@dataclass
class TenantContext:
    """
    Holds tenant information for the current request.

    Mirrors the TenantContext structure from the Go backend:
    /backend/internal/middleware/tenant.go

    Attributes:
        tenant_id: The tenant ID from JWT claims (user's assigned tenant).
        is_super_admin: Whether the user is a super admin.
        effective_tenant_id: The tenant ID to use for queries (may differ
                            for super-admin context switch).
    """

    tenant_id: str
    is_super_admin: bool
    effective_tenant_id: str


def validate_uuid(value: str) -> bool:
    """
    Validate that a string is a valid UUID format.

    Args:
        value: The string to validate.

    Returns:
        True if valid UUID format, False otherwise.
    """
    try:
        UUID(value)
        return True
    except (ValueError, TypeError):
        return False


def create_tenant_context(
    user_claims: UserClaims,
    header_tenant_id: Optional[str] = None
) -> TenantContext:
    """
    Create a TenantContext from user claims and optional header.

    This function replicates the logic from TenantContextMiddleware
    in the Go backend.

    Args:
        user_claims: The authenticated user's claims from JWT.
        header_tenant_id: Optional tenant ID from X-Tenant-Context header.

    Returns:
        TenantContext with appropriate effective_tenant_id.

    Raises:
        TenantContextDeniedError: If non-super-admin tries to switch context.
        InvalidTenantIdError: If header contains invalid tenant ID format.
    """
    # Default effective tenant ID is the user's assigned tenant
    effective_tenant_id = user_claims.tenant_id

    # Check for X-Tenant-Context header (super-admin only)
    if header_tenant_id:
        # Only super-admins can use X-Tenant-Context header
        if not user_claims.is_super_admin:
            raise TenantContextDeniedError()

        # Validate the header tenant ID format
        if not validate_uuid(header_tenant_id):
            raise InvalidTenantIdError(
                "X-Tenant-Context header must contain a valid UUID"
            )

        # Set effective tenant ID to the one from header
        effective_tenant_id = header_tenant_id

    return TenantContext(
        tenant_id=user_claims.tenant_id,
        is_super_admin=user_claims.is_super_admin,
        effective_tenant_id=effective_tenant_id,
    )


def get_tenant_context_from_request(
    user_claims: UserClaims,
    request: Request
) -> TenantContext:
    """
    Extract tenant context from request and user claims.

    Args:
        user_claims: The authenticated user's claims.
        request: The FastAPI request object.

    Returns:
        TenantContext for the current request.

    Raises:
        HTTPException: If tenant context cannot be established.
    """
    header_tenant_id = request.headers.get(TENANT_CONTEXT_HEADER)

    try:
        return create_tenant_context(user_claims, header_tenant_id)
    except TenantContextDeniedError:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={
                "error": "tenant context switch requires super admin privileges",
                "code": "TENANT_CONTEXT_DENIED",
            }
        )
    except InvalidTenantIdError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "error": "invalid X-Tenant-Context header",
                "details": "tenant ID must be a valid UUID",
            }
        )


def require_tenant_context(tenant_ctx: TenantContext) -> None:
    """
    Verify that tenant context is present and valid.

    Args:
        tenant_ctx: The tenant context to validate.

    Raises:
        HTTPException: If tenant context is missing or invalid.
    """
    if not tenant_ctx.effective_tenant_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={
                "error": "tenant ID is required for this operation",
                "code": "TENANT_ID_MISSING",
            }
        )

    if not validate_uuid(tenant_ctx.effective_tenant_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "error": "invalid tenant ID format",
                "code": "INVALID_TENANT_ID",
            }
        )


@dataclass
class RequestContext:
    """
    Combined context for the current request.

    This dataclass combines user claims and tenant context
    into a single object for convenience.
    """

    user: UserClaims
    tenant: TenantContext

    @property
    def user_id(self) -> str:
        """Get the current user's ID."""
        return self.user.user_id

    @property
    def email(self) -> str:
        """Get the current user's email."""
        return self.user.email

    @property
    def role(self) -> str:
        """Get the current user's role."""
        return self.user.role

    @property
    def effective_tenant_id(self) -> str:
        """Get the effective tenant ID for queries."""
        return self.tenant.effective_tenant_id

    @property
    def is_super_admin(self) -> bool:
        """Check if the current user is a super admin."""
        return self.tenant.is_super_admin
